### PR TITLE
Postgre SQL support for DatabaseGpt

### LINKED ISSUE #14: DatabaseGPT for Postgres (and more?)
DatabaseGPT currently supports only scenarios where we have SqlServer databases. It doesn't currently support the case of a different SQL DBMS like Postgres, SQLite, MySql, etc... The major difference between all these DBMS is mainly the dialect of the SQL language used in the database and the required drivers themselves. 

Since a user who is interested in using DatabaseGPT to query a SqlServer might not be interested in having the drivers also for another DBMS, it is preferrable to not include all the possible database drivers in a single package.
I suggest then the refactor of the solution as follows:
- DatabaseGPT: this will be the package that contains all the logic independently from the DBMS. (The actual requests to the OpenAI endpoints should be handled here)
- DatabaseGPT.Core: this will contain the definition of the interfaces and the basic classes required by the database specific implementations. (For example, the definition of an interface IDatabaseProvider)
- DatabaseGPT.SqlServer: this will contain the SqlServer specific logic
- DatabaseGPT.Npgsql: this will contain the Postgres specific logic
- DatabaseGPT.MySql (future)
- DatabaseGPT.SQLite (future)

If the idea is valid and interesting, I can start to design a more detailed refactoring plan and the database-specific implementations for SqlServer and Postgres.

### PR DESCRIPTION
**Pull request by nicolaparo**
Enables the support for Postgre SQL and opens the extension to other DBMSs
Closes #14

### CODE REVIEW

**Review by marcominerva — CHANGES_REQUESTED** [review]
- I have implemented a couple of new features:: https://github.com/marcominerva/DatabaseGPT/pull/17/files. Do you think you can add them too to your pull request?
- Do you mind updating the **README.md** documentation with instruction about the new support for other DBMS?

File: samples/DatabaseGptConsole/Program.cs [hunk 1]
```diff
@@ -28,4 +28,10 @@ static void ConfigureServices(HostBuilderContext context, IServiceCollection ser
 {
     services.AddSingleton<Application>();
     services.AddDatabaseGpt(context.Configuration);
+
+    // For using Postgres
+    // services.AddNpgsqlDatabaseGptProvider(context.Configuration["ConnectionStrings:SqlConnection"]);
+
+    // For using SQL Server
+    services.AddSqlServerDatabaseGptProvider(context.Configuration["ConnectionStrings:SqlConnection"]);
```
[the remark this answers is not in the document]
Comment (nicolaparo) [reply samples/DatabaseGptConsole/Program.cs:36]: Absolutely! I really like the idea. This evening i will implement the feature

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs [hunk 2]
```diff
@@ -42,7 +42,7 @@ public async Task<IDataReader> ExecuteQueryAsync(string query)
         }
         catch (NpgsqlException ex)
         {
-            throw new DatabaseGptException("Something went wrong", ex);
+            throw new DatabaseGptException("An error occured. Se inner exception for details", ex);
```
Comment (marcominerva) [inline src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs:45]: There is a typo. Rewrite the message as it sounds like the typical error messages of this kind, something like `An error occurred while executing the query. See the inner exception for details`.

File: src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs [hunk 3]
```diff
@@ -38,7 +38,7 @@ public async Task<IDataReader> ExecuteQueryAsync(string query)
         }
         catch (SqlException ex)
         {
-            throw new DatabaseGptException("Something went wrong", ex);
+            throw new DatabaseGptException("An error occured. Se inner exception for details", ex);
```
Comment (marcominerva) [inline src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs:41]: There is a typo. Rewrite the message as it sounds like the typical error messages of this kind, something like `An error occurred while executing the query. See the inner exception for details`.

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    private bool disposedValue;
+
+    public NpgsqlDatabaseGptProvider(NpgsqlDatabaseGptProviderConfiguration settings)
+    {
+        this.settings = settings;
+        this.connection = new NpgsqlConnection(settings.ConnectionString);
+        connection.Open();
+    }
+
+    private readonly NpgsqlConnection connection;
```
Comment (marcominerva) [inline src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs:22]: Place this field on top, after the other `readonly` fields.

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    public string Name => "Postgres";
+    public string Language => "Postgres SQL";
```
Comment (marcominerva) [inline src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs:25]: Place these properties on top, after the `private` fields. Place a blank line between the properties.

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+namespace DatabaseGpt.Npgsql;
+
+public class NpgsqlDatabaseGptProvider : IDatabaseGptProvider, IDisposable
+{
+    private readonly NpgsqlDatabaseGptProviderConfiguration settings;
+    private bool disposedValue;
+
+    public NpgsqlDatabaseGptProvider(NpgsqlDatabaseGptProviderConfiguration settings)
+    {
+        this.settings = settings;
```
Comment (marcominerva) [inline src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs:17]: The `settings` argument is used only in the constructor, so you don't need to store it in a private field.

File: src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+namespace DatabaseGpt.SqlServer;
+
+public class SqlServerDatabaseGptProvider : IDatabaseGptProvider, IDisposable
+{
+    private readonly SqlServerDatabaseGptProviderConfiguration settings;
+    private bool disposedValue;
+
+    public SqlServerDatabaseGptProvider(SqlServerDatabaseGptProviderConfiguration settings)
+    {
+        this.settings = settings;
```
Comment (marcominerva) [inline src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs:17]: The `settings` argument is used only in the constructor, so you don't need to store it in a private field.

File: src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        this.connection = new SqlConnection(settings.ConnectionString);
+        connection.Open();
+    }
+
+    private readonly SqlConnection connection;
```
Comment (marcominerva) [inline src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs:22]: Place this field on top, after the other `readonly` fields.

File: src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    public string Name => "SQL Server";
+    public string Language => "T-SQL";
```
Comment (marcominerva) [inline src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs:25]: Place these properties on top, after the `private` fields. Place a blank line between the properties.

File: src/DatabaseGpt/DataAccessLayer/ISqlContext.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/ISqlContext.cs:1]: Remove the unused file.

File: src/DatabaseGpt/DataAccessLayer/SqlContext.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/SqlContext.cs:1]: Remove the unused file.

File: src/DatabaseGpt/DataAccessLayer/SqlContextOptions.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/SqlContextOptions.cs:1]: Remove the unused file.

File: src/DatabaseGpt/DataAccessLayer/SqlContextServiceCollectionExtensions.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/SqlContextServiceCollectionExtensions.cs:1]: Remove the unused file.

File: src/DatabaseGpt/DataAccessLayer/TypeHandlers/DateOnlyTypeHandler.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/TypeHandlers/DateOnlyTypeHandler.cs:1]: This file is necessary to support `date` column for SQL Server, so it must be preserve (and put in the **DatabaseGpt.SqlServer** project).

File: src/DatabaseGpt/DataAccessLayer/TypeHandlers/TimeOnlyTypeHandler.cs
Comment (marcominerva) [inline src/DatabaseGpt/DataAccessLayer/TypeHandlers/TimeOnlyTypeHandler.cs:1]: This file is necessary to support `time` column for SQL Server, so it must be preserve (and put in the **DatabaseGpt.SqlServer** project).

File: samples/DatabaseGptConsole/Program.cs [hunk 10]
```diff
@@ -27,5 +28,15 @@
...
+            .UseConfiguration(context.Configuration)
+            .UseSqlServer(context.Configuration["ConnectionStrings:SqlConnection"]),
+        chatgpt => chatgpt.UseConfiguration(context.Configuration)
+    );
+
+    // For using Postgres
+    // services.AddNpgsqlDatabaseGptProvider(context.Configuration["ConnectionStrings:SqlConnection"]);
+
+    // For using SQL Server
+    // services.AddSqlServerDatabaseGptProvider(context.Configuration["ConnectionStrings:SqlConnection"]);
```
Comment (marcominerva) [inline samples/DatabaseGptConsole/Program.cs:41]: These comments are no longer valid.

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    public string Language => "Postgres SQL";
+
+    public async Task<IEnumerable<string>> GetTablesAsync(IEnumerable<string> excludedTables)
+    {
+        var tables = await connection.QueryAsync<string>("""
+            SELECT TABLE_SCHEMA || '.' || TABLE_NAME AS Tables
+            FROM INFORMATION_SCHEMA.TABLES
+            WHERE TABLE_SCHEMA NOT IN ('pg_catalog', 'information_schema')
+            """);
+        return tables.Where(t => !excludedTables.Contains(t));
```
Comment (marcominerva) [inline src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs:34]: Add a blank line before `return` statement. Add the support for included tables, as in https://github.com/marcominerva/DatabaseGPT/blob/master/src/DatabaseGpt/DatabaseGptClient.cs#L117-L131.

File: src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    public async Task<IEnumerable<string>> GetTablesAsync(IEnumerable<string> excludedTables)
+    {
+        var tables = await connection.QueryAsync<string>("SELECT TABLE_SCHEMA + '.' + TABLE_NAME AS Tables FROM INFORMATION_SCHEMA.TABLES;");
+        return tables.Where(t => !excludedTables.Contains(t));
```
Comment (marcominerva) [inline src/DatabaseGpt.SqlServer/SqlServerDatabaseGptProvider.cs:30]: Add a blank line before `return` statement. Add the support for included tables, as in https://github.com/marcominerva/DatabaseGPT/blob/master/src/DatabaseGpt/DatabaseGptClient.cs#L117-L131.

**Review by marcominerva — CHANGES_REQUESTED** [review]

**Comment by marcominerva** [comment]
@nicolaparo after review the changes, remember to mark them as resolved ;)

File: src/DatabaseGpt/DatabaseGptServiceCollectionExtensions.cs [hunk 13]
```diff
@@ -34,14 +33,44 @@ public static class DatabaseGptServiceCollectionExtensions
...
+
+    //    T? ConfigureAndGet<T>(string sectionName) where T : class
+    //    {
+    //        var section = configuration.GetSection(sectionName);
+    //        var settings = section.Get<T>();
+    //        services.Configure<T>(section);
+
+    //        return settings;
+    //    }
+    //}
```
[the remark this answers is not in the document]
Comment (nicolaparo) [reply src/DatabaseGpt/DatabaseGptServiceCollectionExtensions.cs]: Removed commented code

File: src/DatabaseGpt/DatabaseGptServiceCollectionExtensions.cs [hunk 14]
```diff
@@ -10,20 +10,19 @@ namespace DatabaseGpt;
 
 public static class DatabaseGptServiceCollectionExtensions
 {
-    public static IServiceCollection AddDatabaseGpt(this IServiceCollection services, IConfiguration configuration
-        , ServiceLifetime lifetime = ServiceLifetime.Scoped, string connectionStringName = "SqlConnection", string chatGptSectionName = "ChatGPT")
+    public static IServiceCollection AddDatabaseGpt(this IServiceCollection services, Action<DatabaseGptSettings> buildDatabaseGpt, Action<ChatGptOptionsBuilder> buildChatGpt, ServiceLifetime lifetime = ServiceLifetime.Scoped)
```
[the remark this answers is not in the document]
Comment (nicolaparo) [reply src/DatabaseGpt/DatabaseGptServiceCollectionExtensions.cs]: Yes, it also sounds better.

File: src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs (hunk 2 above)
Comment (nicolaparo) [reply src/DatabaseGpt.Npgsql/NpgsqlDatabaseGptProvider.cs]: Updated the exception message with a more meaningful description.

**Pull request merged**